autoPyTorch/data/utils.py

```diff
@@ -0,0 +1,309 @@
+
+    >>> reduction_mapping = _DtypeReductionMapping()
+    >>>
+    >>> reduction_mapping[np.dtype('float64')] # np.float32
+    >>> np.dtype('float32') in reduction_mapping # True
+    """
+
+    # Information about dtype support
+    _mapping: Dict[type, type] = {
+        np.float32: np.float32,
+        np.float64: np.float32,
+        np.int32: np.int32,
+        np.int64: np.int32
+    }
+
+    # In spite of the names, np.float96 and np.float128
+    # provide only as much precision as np.longdouble,
+    # that is, 80 bits on most x86 machines and 64 bits
+    # in standard Windows builds.
+    if hasattr(np, 'float96'):
+        _mapping[np.float96] = np.float64
+
+    if hasattr(np, 'float128'):
+        _mapping[np.float128] = np.float64
+
+    @classmethod
+    def __getitem__(cls, item: type) -> type:
+        for k, v in cls._mapping.items():
+            if k == item:
+                return v
+        raise KeyError(item)
+
+    @classmethod
+    def __iter__(cls) -> Iterator[type]:
+        return iter(cls._mapping.keys())
+
+    @classmethod
+    def __len__(cls) -> int:
+        return len(cls._mapping)
+
+
+reduction_mapping = _DtypeReductionMapping()
+supported_precision_reductions = list(reduction_mapping)
+
+
+def reduce_precision(
+    X: DatasetCompressionInputType
+) -> Tuple[DatasetCompressionInputType, DatasetDTypeContainerType, DatasetDTypeContainerType]:
+    """ Reduces the precision of a dataset containing floats or ints
+
+    Note:
+        For dataframe, the column's precision is reduced using pd.to_numeric.
+
+    Args:
+        X:  DatasetCompressionInputType
+            The data to reduce precision of.
+
+    Returns:
+        Tuple[DatasetCompressionInputType, DatasetDTypeContainerType, DatasetDTypeContainerType]
+            Returns the reduced data X along with the dtypes it and the dtypes it was reduced to.
+    """
+    precision: Optional[DatasetDTypeContainerType] = None
+    if isinstance(X, np.ndarray) or issparse(X):
+        dtypes = X.dtype
+        if X.dtype not in supported_precision_reductions:
+            raise ValueError(f"X.dtype = {X.dtype} not equal to any supported"
+                             f" {supported_precision_reductions}")
+        precision = reduction_mapping[X.dtype]
+        X = X.astype(precision)
+    elif hasattr(X, 'iloc'):
+        dtypes = dict(X.dtypes)
+
+        integer_columns = []
+        float_columns = []
+
+        for col, dtype in dtypes.items():
+            if is_numeric_dtype(dtype):
+                if is_float_dtype(dtype):
+                    float_columns.append(col)
+                else:
+                    integer_columns.append(col)
+
+        if len(integer_columns) > 0:
+            X[integer_columns] = X[integer_columns].apply(lambda column: pd.to_numeric(column, downcast='integer'))
+        if len(float_columns) > 0:
+            X[float_columns] = X[float_columns].apply(lambda column: pd.to_numeric(column, downcast='float'))
```

Comment:
```suggestion
        integer_columns = X.columns[X.dtypes == int]
        float_columns = X.columns[X.dtypes == float]
        X[integer_columns] = X[integer_columns].apply(lambda column: pd.to_numeric(column, downcast='integer'))
        X[float_columns] = X[float_columns].apply(lambda column: pd.to_numeric(column, downcast='float'))
```

Reply:
no, actually pandas has a dtype `Float64DType` which is present when there are `nans`. I couldnt find another way to check for this dtype.

Reply:
Let me check then...

Reply:
```python
        col_names = X.dtypes.index
        float_cols = col_names[[dt in FLOAT_DTYPES for dt in X.dtypes.values]]
        int_cols = col_names[[dt in INT_DTYPES for dt in X.dtypes.values]]
        X[int_cols] = X[int_cols].apply(lambda column: pd.to_numeric(column, downcast='integer'))
        X[float_cols] = X[float_cols].apply(lambda column: pd.to_numeric(column, downcast='float'))
```